File: Stacks_Queues/C/stacks.c
/*
this implementation has a fixed size, if you want a dynamic size stack you can use linked list or dynamic array.
Time complexity for all operations is O(1)
*/
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

struct Stack {
    int top;
    int data[MAX_SIZE];
};

void init(struct Stack* stack) {
    stack->top = -1;
}

int is_empty(struct Stack* stack) {
    return stack->top == -1;
}

int is_full(struct Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

void push(struct Stack* stack, int item) {
    if (is_full(stack)) {
        printf("Error: Stack is full\n");
        return;
    }
    stack->data[++stack->top] = item;
}

int pop(struct Stack* stack) {
    if (is_empty(stack)) {
        printf("Error: Stack is empty\n");
        return -1;
    }
    return stack->data[stack->top--];
}

int peek(struct Stack* stack) {
    if (is_empty(stack)) {
        printf("Error: Stack is empty\n");
        return -1;
    }
    return stack->data[stack->top];
}

int main() {
    struct Stack stack;
    init(&stack);
    push(&stack, 1);
    push(&stack, 2);
    push(&stack, 3);
    printf("%d\n", peek(&stack)); // Output: 3
    printf("%d\n", pop(&stack)); // Output: 3
    printf("%d\n", peek(&stack)); // Output: 2
    printf("%d\n", pop(&stack)); // Output: 2
    printf("%d\n", peek(&stack)); // Output: 1
    printf("%d\n", pop(&stack)); // Output: 1
    printf("%d\n", is_empty(&stack)); // Output: 1 (true)
    return 0;
}

File: Stacks_Queues/Python/stacks.py
"""
this implementation has a time complexity of O(1) for push, pop and top operations
if you use a linked list or dynamic array the time complexity will be O(n)
"""
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()

    def peek(self):
        if not self.is_empty():
            return self.items[-1]

    def is_empty(self):
        return len(self.items) == 0

stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)
print(stack.peek()) # Output: 3
stack.pop()
print(stack.peek()) # Output: 2
stack.pop()
print(stack.peek()) # Output: 1
stack.pop()
print(stack.is_empty()) # Output: True
